src/d3a/models/strategy/external_strategies/__init__.py

```diff
@@ -444,38 +442,101 @@ def _set_energy_forecast(self, payload):
                 IncomingRequest("set_energy_forecast", arguments,
                                 energy_forecast_response_channel))
 
-    def _set_energy_forecast_impl(self, arguments, response_channel):
+    def _set_energy_measurement(self, payload: Dict) -> None:
+        """Callback of set_energy_measurement redis endpoint."""
+        transaction_id = self._get_transaction_id(payload)
+        energy_measurement_response_channel = \
+            f"{self.channel_prefix}/response/set_energy_measurement"
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {"energy_measurement", "transaction_id"}
+        except Exception:
+            logging.exception(
+                f"Incorrect _set_energy_measurement request. Payload: {payload}.")
+            self.redis.publish_json(
+                energy_measurement_response_channel,
+                {"command": "set_energy_measurement",
+                 "error": "Incorrect _set_energy_measurement request. "
+                          "Available parameters: (energy_measurement).",
+                 "transaction_id": transaction_id})
+        else:
+            self.pending_requests.append(
+                IncomingRequest("set_energy_measurement", arguments,
+                                energy_measurement_response_channel))
+
+    @staticmethod
+    def _validate_values_positive_in_profile(profile: Dict) -> None:
+        """Validate whether all values are positive in a profile."""
+        for time_str, energy in profile.items():
+            assert energy >= 0.0, f"Energy is not positive for time stamp {time_str}."
```

Comment:
Nit: if this is dealing with external input (e.g. user input or client) it would be better to raise an exception instead of using assertions.

Reply:
This function is called in a try except block which will deal with sending the error back to the client.
You prepended `Nit:`, so I guess that I could skip it right?

Reply:
TL;DR: This is not blocking so you can skip it. But still quite important and it would be good to keep it in mind for the future.

The try/except block could still stay where it is. Assertions are for errors that should never happen in production, they're meant for developers (so that they know they have a bug in their code) and could/should fail during local development with a bad crash. Also, assertions can be disabled globally with a specific python flag.
Exceptions (with try/except) are errors that are expected, they're meant for production code and should be caught to avoid crashes and provide users with meaningful error messages.
In this case, we're validating the client input, so we expect it to raise exceptions in some cases. It's a normal thing and it does not depend on a bug in the code. That's why I'm suggesting to use exceptions.

Reply:
OK, will add an exception here.